weave/trace/serialization/README.md

```diff
@@ -42,15 +42,22 @@ NOTE: Not all serialization is reversible, which can be surprising and frustrati
 
 ### 2. Custom Object Serialization (`custom_objs.py` and `serializer.py`)
 
-NOTE: Custom object serialization is WIP and subject to change!
+NOTE: Custom object serialization is WIP and subject to change! In particular, there are some asymmetries and other issues that we should fix:
+
+1. Encoding requires a WeaveClient while decoding requires a TraceServerInterface. Encoding is more complicated because of asynchronous file creation and caching in the client. A client object contains a reference to a TraceServerInterface so ideally we could use a WeaveClient object for both encoding and decoding.
+2. We are currently passing a project_id as a separate argument - a client has entity/project information internally so we shouldn't need to pass it separately.
+3. to_json and from_json should not need to know about inline vs. file-backed custom objects, the logic for things like file creation should be moved from `serializer.py` to `custom_objs.py` - this is complicated by #1 above. If we fix this we could expose a single `decode_custom_obj` method mirroring the single encode method.
 
 The entrypoint for custom object serialization is `weave/trace/serialization/custom_objs.py`. This file primarily contains:
 
-1. `encode_custom_obj()`, which encodes custom objects using registered serializers
-2. `decode_custom_obj()`, which decodes custom objects using the appropriate serializer
+1. `encode_custom_obj()`, which encodes custom objects using registered serializers (both inline and file-based)
+2. `decode_custom_inline_obj()`, which decodes custom inline objects using the appropriate serializer
+3. `decode_custom_files_obj()`, which decodes custom file-based objects using the appropriate serializer
 
 #### Custom Object Serialization Flow
 
+This diagram shows the flow for file-based custom objects. Inline objects are similar but simpler.
```

Comment:
Can you show the inline diagram too?

Reply:
I'll add one, but I think they are a little confusing (e.g. it looks like the decode is calling the from_json when it is the other way around)